File: 0525-contiguous-array/0525-contiguous-array.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function(nums) {
    let mp = new Map();
    mp.set(0, -1);
    let sum = 0, maxLen = 0;
    let n = nums.length;

    for(let i = 0; i<n; i++) {
        sum += nums[i] == 0 ? -1 : 1;
        if(mp.has(sum)) {
            maxLen = Math.max(maxLen, i - mp.get(sum));
        }
        else mp.set(sum, i);
    }

    return maxLen;
};

File: 0525-contiguous-array/0525-contiguous-array.cpp
class Solution {
public:
    int findMaxLength(vector<int>& nums) {
        int n = nums.size();
        unordered_map<int,int> mp;
        mp[0] = -1;
        int sum = 0, maxlen = 0;

        for(int i = 0; i<n; i++) {
            sum += nums[i] == 0 ? -1 : 1;
            if(mp.find(sum) != mp.end()) {
                maxlen = max(maxlen, i - mp[sum]);
            }
            else mp[sum] = i;
        }

        return maxlen;
    }
};